api/proto/objects/revset.proto

```diff
@@ -0,0 +1,72 @@
+syntax = "proto3";
+
+package jj_api.objects;
+
+message Revset {
```

Comment:
I'm not a huge fan of that idea. The user input is just a string, so doesn't need its own type, and I think the term `BuiltinRevset` just adds additional confusion to a user of the API.

@martinvonz WDYT?

Reply:
If we allow revsets strings here, then we presumably allow nesting revset strings inside other revset nodes (e.g. `intersection=[change_id="xyz", revset="main.."]`). That seems reasonable and useful for a UI.

Reply:
> I'm not a huge fan of that idea. The user input is just a string, so doesn't need its own type, and I think the term `BuiltinRevset` just adds additional confusion to a user of the API.

Then how do custom downstreams implement builtins? Imagine if Google implements a `commit_visible_to(user, group)` for Piper, how would you access it in the implementation? Anyways that was my thought.

Reply:
My assumption was that the custom downstream would just use `Revset(revset="commit_visible_to(user, group)")`. I'm also not sure I understand how your proposal would solve that problem.

Alternatively, we could just add a type for a custom function call.
```
message Arg {
  oneof {
    Revset revset = 1;
    string literal = 2;
  }
}

message FunctionCall {
  string function_name = 2;
  repeated Arg args;

message Revset {
  oneof kind {
    // existing options
    FunctionCall function_call;
  }
}
```

Reply:
> My assumption was that the custom downstream would just use `Revset(revset="commit_visible_to(user, group)")`.

Yeah, that definitely works. 

> I'm also not sure I understand how your proposal would solve that problem.

It would allow downstreams to add it to their "BuiltinRevset" enumeration, although this could also be problematic if it leads to tag overlaps. 

> Alternatively, we could just add a type for a custom function call.

I like that solution, but it is very low-level and I'm not sure if that aligns with the stated goal.